crates/astria-conductor/src/celestia/reconstruct.rs

```diff
@@ -0,0 +1,201 @@
+///    `rollup_id`:
+///    - if they contained `rollup_id` they are dropped (as they should have had a matching blob but
+///      none was found).
+///    - if they did not contain `rollup_id` a Sequencer block is reconstructed.
+pub(super) fn reconstruct_blocks_from_verified_blobs(
+    verified_blobs: VerifiedBlobs,
+    rollup_id: RollupId,
+) -> Vec<ReconstructedBlock> {
+    let (celestia_height, header_blobs, rollup_blobs) = verified_blobs.into_parts();
+
+    let mut headers = BlockHashAndHeightToHeader::from_header_blobs(header_blobs);
+
+    let mut reconstructed_blocks = Vec::new();
+    for rollup in rollup_blobs {
+        if let Some(header_blob) = headers.remove_header_blob_matching_rollup_blob(&rollup) {
+            reconstructed_blocks.push(ReconstructedBlock {
+                celestia_height,
+                block_hash: header_blob.block_hash(),
+                header: header_blob.into_unchecked().header,
+                transactions: rollup.into_unchecked().transactions,
+            });
+        } else {
+            let reason = if headers.any_headers_with_block_hash(rollup.sequencer_block_hash()) {
+                "sequencer header blobs with the same block hash as the rollup blob found, but the \
+                 rollup's Merkle proof did not lead any Merkle roots"
+            } else {
+                "no sequencer header blob matching the rollup blob's block hash found"
+            };
+            info!(
+                block_hash = %base64(&rollup.sequencer_block_hash()),
+                reason,
+                "dropping rollup blob",
+            );
+        }
+    }
+
+    let remaining_headers = headers.map;
+
+    for header_blob in remaining_headers.into_values() {
+        if header_blob.contains_rollup_id(rollup_id) {
+            warn!(
+                block_hash = %base64(&header_blob.block_hash()),
+                "sequencer header blob contains the target rollup ID, but no matching rollup blob was found; the blob",
+            );
+        } else {
+            reconstructed_blocks.push(ReconstructedBlock {
+                celestia_height,
+                block_hash: header_blob.block_hash(),
+                header: header_blob.into_unchecked().header,
+                transactions: vec![],
+            });
+        }
+    }
+    reconstructed_blocks
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+struct Key {
+    block_hash: [u8; 32],
+    height: SequencerHeight,
+}
+
+impl Key {
+    fn from_header_blob(val: &CelestiaSequencerBlob) -> Self {
+        Self {
+            block_hash: val.block_hash(),
+            height: val.height(),
+        }
+    }
+}
+
+struct BlockHashAndHeightToHeader {
+    // The sequencer header blobs grouped by (block_hash, sequencer_height)
+    map: HashMap<Key, CelestiaSequencerBlob>,
+
+    // A reverse index of block_hash -> [(block_hash, sequencer_height)]
+    reverse_index: HashMap<[u8; 32], Vec<Key>>,
```

Comment:
why is the value a vec? wouldn't a specific block hash always be for a specific sequencer height?

Reply:
The idea was that there could be multiple `(block_hash, sequencer_height)` pairs with the same `block_hash` but different `sequencer_height`s.

Reply:
The verify phase is now constructing a map `block_hash -> CelestiaSequencerBlob` so that I could simplify this code.